README.md
# NFS Ganesha server and external provisioner

`nfs-ganesha-server-and-external-provisioner` is an out-of-tree dynamic provisioner for Kubernetes 1.14+. You can use it to quickly & easily deploy shared storage that works almost anywhere. Or it can help you write your own out-of-tree dynamic provisioner by serving as an example implementation of the requirements detailed in [the proposal](https://github.com/kubernetes/kubernetes/pull/30285). 

It works just like in-tree dynamic provisioners: a `StorageClass` object can specify an instance of `nfs-ganesha-server-and-external-provisioner` to be its `provisioner` like it specifies in-tree provisioners such as GCE or AWS. Then, the instance of nfs-ganesha-server-and-external-provisioner will watch for `PersistentVolumeClaims` that ask for the `StorageClass` and automatically create NFS-backed `PersistentVolumes` for them. For more information on how dynamic provisioning works, see [the docs](http://kubernetes.io/docs/user-guide/persistent-volumes/) or [this blog post](http://blog.kubernetes.io/2016/10/dynamic-provisioning-and-storage-in-kubernetes.html).

Note: This repository was migrated from https://github.com/kubernetes-incubator/external-storage/tree/HEAD/nfs. Some of the following instructions will be updated once the build and release automtion is setup. To test container image built from this repository, you will have to build and push the nfs-provisioner image using the following instructions.

```sh
make build
make container
# `nfs-provisioner:latest` will be created. 
# To upload this to your customer registry, say `gcr.io/myorg`, you can use
# docker tag nfs-provisioner:latest gcr.io/myorg/nfs-provisioner:latest
# docker push gcr.io/myorg/nfs-provisioner:latest
```

## Quickstart

Choose some volume for your `nfs-ganesha-server-and-external-provisioner` instance to store its state & data in and mount the volume at `/export` in [deploy/kubernetes/deployment.yaml](./deploy/kubernetes/deployment.yaml). It doesn't have to be a `hostPath` volume, it can e.g. be a PVC. Note that the volume must have a [supported file system](https://github.com/nfs-ganesha/nfs-ganesha/wiki/Fsalsupport#vfs) on it: any local filesystem on Linux is supported & NFS is not supported.

```yaml
...
  volumeMounts:
    - name: export-volume
      mountPath: /export
volumes:
  - name: export-volume
    hostPath:
      path: /tmp/nfs-provisioner
...
```

Choose a `provisioner` name for a `StorageClass` to specify and set it in `deploy/kubernetes/deployment.yaml`
```yaml
...
args:
  - "-provisioner=example.com/nfs"
...
```

Create the deployment.
```console
$ kubectl create -f deploy/kubernetes/deployment.yaml
serviceaccount/nfs-provisioner created
service "nfs-provisioner" created
deployment "nfs-provisioner" created
```

Create `ClusterRole`, `ClusterRoleBinding`, `Role` and `RoleBinding` (this is necessary if you use RBAC authorization on your cluster, which is the default for newer kubernetes versions).
```console
$ kubectl create -f deploy/kubernetes/rbac.yaml
clusterrole.rbac.authorization.k8s.io/nfs-provisioner-runner created
clusterrolebinding.rbac.authorization.k8s.io/run-nfs-provisioner created
role.rbac.authorization.k8s.io/leader-locking-nfs-provisioner created
rolebinding.rbac.authorization.k8s.io/leader-locking-nfs-provisioner created
```

Create a `StorageClass` named "example-nfs" with `provisioner: example.com/nfs`.
```console
$ kubectl create -f deploy/kubernetes/class.yaml
storageclass "example-nfs" created
```

Create a `PersistentVolumeClaim` with `storageClassName: example-nfs`.
```console
$ kubectl create -f deploy/kubernetes/claim.yaml
persistentvolumeclaim "nfs" created
```

A `PersistentVolume` is provisioned for the `PersistentVolumeClaim`. Now the claim can be consumed by some pod(s) and the backing NFS storage read from or written to.
```console
$ kubectl get pv
NAME                                       CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS      CLAIM         REASON    AGE
pvc-dce84888-7a9d-11e6-b1ee-5254001e0c1b   1Mi        RWX           Delete          Bound       default/nfs             23s
```

Deleting the `PersistentVolumeClaim` will cause the provisioner to delete the `PersistentVolume` and its data.

Deleting the provisioner deployment will cause any outstanding `PersistentVolumes` to become unusable for as long as the provisioner is gone.

## Running

To deploy `nfs-ganesha-server-and-external-provisioner` on a Kubernetes cluster see [Deployment](docs/deployment.md).

To use `nfs-ganesha-server-and-external-provisioner` once it is deployed see [Usage](docs/usage.md).

## [Changelog](CHANGELOG.md)

Releases done here in external-storage will not have corresponding git tags (external-storage's git tags are reserved for versioning the library), so to keep track of releases check this README, the [changelog](CHANGELOG.md), or [GCR](https://gcr.io/k8s-staging-sig-storage/nfs-provisioner)

## Writing your own

Go [here](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/tree/HEAD/examples/hostpath-provisioner) for an example of how to write your own out-of-tree dynamic provisioner.

## Roadmap

The source code in this repository was migrated from [kubernetes-incubator/external-storage](https://github.com/kubernetes-incubator/external-storage/tree/HEAD/nfs). We are yet to complete the following migration tasks.
- Update e2e tests
- Automate building container images to the new registry
- Update helm chart

This is still alpha/experimental and will change to reflect the [out-of-tree dynamic provisioner proposal](https://github.com/kubernetes/kubernetes/pull/30285)

## Community, discussion, contribution, and support

Learn how to engage with the Kubernetes community on the [community page](http://kubernetes.io/community/).

You can reach the maintainers of this project at:

- [Slack](https://kubernetes.slack.com/messages/sig-storage)
- [Mailing List](https://groups.google.com/forum/#!forum/kubernetes-sig-storage)

### Code of conduct

Participation in the Kubernetes community is governed by the [Kubernetes Code of Conduct](code-of-conduct.md).


CHANGELOG.md
# v1.6.0
- Rebase on Fedora 36
- Upgrade Ganesha to 4.0.8
- Switch to new repo for dbus

# v1.0.8
- Add mountOptions StorageClass parameter (#84) (see [Usage](./docs/usage.md) for complete SC parameter info)
- Replace root-squash argument with a rootSquash SC parameter (#86) (see [Usage](./docs/usage.md) for complete SC parameter info)
	- If the root-squash argument is specified, the provisioner will fail to start; please if you're using it, convert to the SC parameter before updating!
- Watch for unexpected stop of ganesha.nfsd and restart if seen (#98). This is a simple health check that mitigates NFS ganesha crashes which are under investigation (but probably out of the provisioner's control to prevent at the moment).

# v1.0.7
- Set a high limit for maximum number of files Ganesha may have open (setrlimit RLIMIT_NOFILE) -- this requires the additional SYS_RESOURCE capability, if not available the provisioner will still start but with a warning

# v1.0.6
- Reduce image size by a lot

# v1.0.5
- Add compatibility with kubernetes v1.6.x (using lib v2.0.x)

# v1.0.4
- Add `server-hostname` flag

# Rename kubernetes-incubator/nfs-provisioner to kubernetes-incubator/external-storage
- The previous releases were done when the repo was named nfs-provisioner: http://github.com/kubernetes-incubator/nfs-provisioner/releases. Newer releases done here in external-storage will *not* have corresponding git tags (external-storage's git tags are reserved for versioning the library), so to keep track of releases check this changelog, the [README](README.md), or [Quay](https://quay.io/repository/kubernetes_incubator/nfs-provisioner)

# v1.0.3
- Fix inability to export NFS shares ("1 validation errors in block FSAL") when using Docker's overlay storage driver (CoreOS/container linux, GCE) by patching Ganesha to use device number as fsid. (#63)
- Adds configurable number of retries on failed Provisioner operations. Configurable as an argument to `NewProvisionController`. nfs-provisioner defaults to 10 retries unless the new flag/argument is used. (#65)

# v1.0.2
- Usage demo & how-to for writing your own external PV provisioner added here https://github.com/kubernetes-incubator/nfs-provisioner/tree/HEAD/demo
- Change behaviour for getting NFS server IP from env vars (node, service) in case POD_IP env var is not set when needed. Use `hostname -i` as a fallback only for when running out-of-cluster (#52)
- Pass whole PVC object from controller to `Provision` as part of `VolumeOptions`, like upstream (#48)
- Filter out controller's self-generated race-to-lock leader election PVC updates from being seen as forced resync PVC updates (#58) 
- Fix controller's event watching for ending race-to-lock leader elections early. Now correctly discover the latest `ProvisionFailed`/`ProvisionSucceeded` events on a claim (#59)

# v1.0.1
- Add rootsquash flag for enabling/disabling rootsquash https://github.com/kubernetes-incubator/nfs-provisioner/pull/40

# v1.0.0
- Automatically create NFS PVs for any user-defined Storage Class of PVCs, backed by a containerized NFS server that creates & exports shares from some user-defined mounted storage
- Support multiple ways to run:
  - standalone Pod, e.g. for easy dynamically provisioned scratch space
  - stateful app, either as a StatefulSet or Deployment of 1 replica: the NFS server will survive restarts and its provisioned PVs can be backed by some mounted persistent storage e.g. a hostPath or one big PV
  - DaemonSet, where each node runs the NFS server to expose its hostPath storage
  - Docker container or binary outside of Kube
- Race-to-lock PVCs: when multiple instances are running & serving the same PVCs, only one attempts to provision for a PVC at a time
- Optionally exponentially backoff from calls to Provision() and Delete()
- Optionally set per-PV filesystem quotas: based on XFS project-level quotas and available only when running outside of Kubernetes (pending mount option support in Kube)

Docker image:
`quay.io/kubernetes_incubator/nfs-provisioner:v1.0.0`


test/e2e/test.sh
#!/bin/bash

# Copyright 2018 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Vendoring the e2e framework is too hard. Download kubernetes source, patch
# our tests on top of it, build and run from there.

KUBE_VERSION=1.11.0
TEST_DIR=$GOPATH/src/github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner/test/e2e

GOPATH=$TEST_DIR

# Download kubernetes source
if [ ! -e "$GOPATH/src/k8s.io/kubernetes" ]; then
  mkdir -p $GOPATH/src/k8s.io
  curl -L https://github.com/kubernetes/kubernetes/archive/v${KUBE_VERSION}.tar.gz | tar xz -C $TEST_DIR/src/k8s.io/
  rm -rf $GOPATH/src/k8s.io/kubernetes
  mv $GOPATH/src/k8s.io/kubernetes-$KUBE_VERSION $GOPATH/src/k8s.io/kubernetes
fi

cd $GOPATH/src/k8s.io/kubernetes

# Clean some unneeded sources
find ./test/e2e -maxdepth 1 -type d ! -name 'e2e' ! -name 'framework' ! -name 'manifest' ! -name 'common' ! -name 'generated' ! -name 'testing-manifests' ! -name 'perftype' -exec rm -r {} +
find ./test/e2e -maxdepth 1 -type f \( -name 'examples.go' -o -name 'gke_local_ssd.go' -o -name 'gke_node_pools.go' \) -delete
find ./test/e2e/testing-manifests -maxdepth 1 ! -name 'testing-manifests' ! -name 'BUILD' -exec rm -r {} +

# Copy our sources
mkdir ./test/e2e/storage
cp $TEST_DIR/nfs.go ./test/e2e/storage/
rm ./test/e2e/e2e_test.go
cp $TEST_DIR/e2e_test.go ./test/e2e/
cp -r $TEST_DIR/testing-manifests/* ./test/e2e/testing-manifests

# Build e2e.test
./build/run.sh make KUBE_BUILD_PLATFORMS=linux/amd64 WHAT=test/e2e/e2e.test &> /dev/null

# Download kubectl to _output directory
if [ ! -e "$HOME/bin/kubectl" ]; then
  curl -o $HOME/bin/kubectl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl
  chmod +x $HOME/bin/kubectl
fi

# Run tests assuming local cluster i.e. one started with hack/local-up-cluster.sh
./_output/dockerized/bin/linux/amd64/e2e.test --provider=local --ginkgo.focus=external-storage --kubeconfig=$HOME/.kube/config


.prow.sh
#! /bin/bash

# Copyright 2020 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# A Prow job can override these defaults, but this shouldn't be necessary.

# At the moment, only linux builds are supported by the ./Dockerfile. 
: ${CSI_PROW_BUILD_PLATFORMS:="linux amd64; linux arm64 -arm64; linux ppc64le -ppc64le; linux s390x -s390x"}

# Only these tests make sense until we can integrate k/k
# e2es.
: ${CSI_PROW_TESTS:="unit"}

. release-tools/prow.sh

main


.cloudbuild.sh
#! /bin/bash

# Copyright 2020 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# At the moment, only linux builds are supported by the ./Dockerfile.
: ${CSI_PROW_BUILD_PLATFORMS:="linux amd64; linux arm64 -arm64; linux ppc64le -ppc64le; linux s390x -s390x"}

# shellcheck disable=SC1091
. release-tools/cloudbuild.sh


pkg/volume/util.go
/*
Copyright 2016 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package volume

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
)

const (
	// ExportID = 152 is being reserved
	// so that 152.152 is not used as filesystem_id in nfs-ganesha export configuration
	// 152.152 is the default pseudo root filesystem ID
	// Ref: https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner/issues/7
	ReservedExportID = 152
)

// generateID generates a unique exportID to assign an export
func generateID(mutex *sync.Mutex, ids map[uint16]bool) uint16 {
	mutex.Lock()
	id := uint16(1)
	for ; id <= math.MaxUint16; id++ {
		if _, ok := ids[id]; !ok {
			if id == ReservedExportID {
				continue
			}
			break
		}
	}
	ids[id] = true
	mutex.Unlock()
	return id
}

func deleteID(mutex *sync.Mutex, ids map[uint16]bool, id uint16) {
	mutex.Lock()
	delete(ids, id)
	mutex.Unlock()
}

// getExistingIDs populates a map with existing ids found in the given config
// file using the given regexp. Regexp must have a "digits" submatch.
func getExistingIDs(config string, re *regexp.Regexp) (map[uint16]bool, error) {
	ids := map[uint16]bool{}

	digitsRe := "([0-9]+)"
	if !strings.Contains(re.String(), digitsRe) {
		return ids, fmt.Errorf("regexp %s doesn't contain digits submatch %s", re.String(), digitsRe)
	}

	read, err := ioutil.ReadFile(config)
	if err != nil {
		return ids, err
	}

	allMatches := re.FindAllSubmatch(read, -1)
	for _, match := range allMatches {
		digits := match[1]
		if id, err := strconv.ParseUint(string(digits), 10, 16); err == nil {
			ids[uint16(id)] = true
		}
	}

	return ids, nil
}

func addToFile(mutex *sync.Mutex, path string, toAdd string) error {
	mutex.Lock()

	file, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0600)
	if err != nil {
		mutex.Unlock()
		return err
	}
	defer file.Close()

	if _, err = file.WriteString(toAdd); err != nil {
		mutex.Unlock()
		return err
	}
	file.Sync()

	mutex.Unlock()
	return nil
}

func removeFromFile(mutex *sync.Mutex, path string, toRemove string) error {
	mutex.Lock()

	read, err := ioutil.ReadFile(path)
	if err != nil {
		mutex.Unlock()
		return err
	}

	removed := strings.Replace(string(read), toRemove, "", -1)
	err = ioutil.WriteFile(path, []byte(removed), 0)
	if err != nil {
		mutex.Unlock()
		return err
	}

	mutex.Unlock()
	return nil
}


pkg/server/server.go
/*
Copyright 2016 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package server

import (
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"syscall"

	"github.com/golang/glog"
)

var defaultGaneshaConfigContents = []byte(`
###################################################
#
# EXPORT
#
# To function, all that is required is an EXPORT
#
# Define the absolute minimal export
#
###################################################

EXPORT
{
	# Export Id (mandatory, each EXPORT must have a unique Export_Id)
	Export_Id = 0;

	# Exported path (mandatory)
	Path = /nonexistent;

	# Pseudo Path (required for NFS v4)
	Pseudo = /nonexistent;

	# Required for access (default is None)
	# Could use CLIENT blocks instead
	Access_Type = RW;

	# Exporting FSAL
	FSAL {
		Name = VFS;
	}
}

NFS_Core_Param
{
	MNT_Port = 20048;
	NLM_Port = 32803;
	fsid_device = true;
}

NFSV4
{
	Grace_Period = 90;
	Only_Numeric_Owners = false;
}
`)

// Setup sets up various prerequisites and settings for the server. If an error
// is encountered at any point it returns it instantly
func Setup(ganeshaConfig string, gracePeriod uint, fsidDevice, disableIDMapping bool) error {
	// Start rpcbind if it is not started yet
	cmd := exec.Command("/usr/sbin/rpcinfo", "127.0.0.1")
	if err := cmd.Run(); err != nil {
		cmd = exec.Command("/usr/sbin/rpcbind", "-w")
		if out, err := cmd.CombinedOutput(); err != nil {
			return fmt.Errorf("Starting rpcbind failed with error: %v, output: %s", err, out)
		}
	}

	cmd = exec.Command("/usr/sbin/rpc.statd", "--port", "662")
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("rpc.statd failed with error: %v, output: %s", err, out)
	}

	// Start dbus, needed for ganesha dynamic exports
	cmd = exec.Command("dbus-daemon", "--system", "--nopidfile")
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("dbus-daemon failed with error: %v, output: %s", err, out)
	}

	err := setRlimitNOFILE()
	if err != nil {
		glog.Warningf("Error setting RLIMIT_NOFILE, there may be 'Too many open files' errors later: %v", err)
	}

	// Use defaultGaneshaConfigContents if the ganeshaConfig doesn't exist yet
	if _, err = os.Stat(ganeshaConfig); os.IsNotExist(err) {
		err = ioutil.WriteFile(ganeshaConfig, defaultGaneshaConfigContents, 0600)
		if err != nil {
			return fmt.Errorf("error writing ganesha config %s: %v", ganeshaConfig, err)
		}
	}
	err = setGracePeriod(ganeshaConfig, gracePeriod)
	if err != nil {
		return fmt.Errorf("error setting grace period to ganesha config: %v", err)
	}
	err = setNumericOwners(ganeshaConfig, disableIDMapping)
	if err != nil {
		return fmt.Errorf("error setting grace period to ganesha config: %v", err)
	}
	err = setFsidDevice(ganeshaConfig, fsidDevice)
	if err != nil {
		return fmt.Errorf("error setting fsid device to ganesha config: %v", err)
	}
	err = setNlmPort(ganeshaConfig)
	if err != nil {
		return fmt.Errorf("error setting NLM port to ganesha config: %v", err)
	}

	return nil
}

// Run : run the NFS server in the foreground until it exits
// Ideally, it should never exit when run in foreground mode
// We force foreground to allow the provisioner process to restart
// the server if it crashes - daemonization prevents us from using Wait()
// for this purpose
func Run(ganeshaLog, ganeshaPid, ganeshaConfig string) error {
	// Start ganesha.nfsd
	glog.Infof("Running NFS server!")
	cmd := exec.Command("ganesha.nfsd", "-F", "-L", ganeshaLog, "-p", ganeshaPid, "-f", ganeshaConfig)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("ganesha.nfsd failed with error: %v, output: %s", err, out)
	}

	return nil
}

func setRlimitNOFILE() error {
	var rlimit syscall.Rlimit
	err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit)
	if err != nil {
		return fmt.Errorf("error getting RLIMIT_NOFILE: %v", err)
	}
	glog.Infof("starting RLIMIT_NOFILE rlimit.Cur %d, rlimit.Max %d", rlimit.Cur, rlimit.Max)
	rlimit.Max = 1024 * 1024
	rlimit.Cur = 1024 * 1024
	err = syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rlimit)
	if err != nil {
		return err
	}
	err = syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit)
	if err != nil {
		return fmt.Errorf("error getting RLIMIT_NOFILE: %v", err)
	}
	glog.Infof("ending RLIMIT_NOFILE rlimit.Cur %d, rlimit.Max %d", rlimit.Cur, rlimit.Max)
	return nil
}

func setFsidDevice(ganeshaConfig string, fsidDevice bool) error {
	newLine := fmt.Sprintf("fsid_device = %t;", fsidDevice)

	re := regexp.MustCompile("fsid_device = (true|false);")

	read, err := ioutil.ReadFile(ganeshaConfig)
	if err != nil {
		return err
	}

	oldLine := re.Find(read)

	if oldLine == nil {
		// fsid_device line not there, append it after MNT_Port
		re := regexp.MustCompile("MNT_Port = 20048;")

		mntPort := re.Find(read)

		block := "MNT_Port = 20048;\n" +
			"\t" + newLine

		replaced := strings.Replace(string(read), string(mntPort), block, -1)
		err = ioutil.WriteFile(ganeshaConfig, []byte(replaced), 0)
		if err != nil {
			return err
		}
	} else {
		// fsid_device there, just replace it
		replaced := strings.Replace(string(read), string(oldLine), newLine, -1)
		err = ioutil.WriteFile(ganeshaConfig, []byte(replaced), 0)
		if err != nil {
			return err
		}
	}

	return nil
}

func setNlmPort(ganeshaConfig string) error {
	newLine := "NLM_Port = 32803;"

	re := regexp.MustCompile("NLM_Port = ([0-9]+);")

	read, err := ioutil.ReadFile(ganeshaConfig)
	if err != nil {
		return err
	}

	oldLine := re.Find(read)

	if oldLine == nil {
		// fsid_device line not there, append it after MNT_Port
		re := regexp.MustCompile("MNT_Port = 20048;")

		mntPort := re.Find(read)

		block := "MNT_Port = 20048;\n" +
			"\t" + newLine

		replaced := strings.Replace(string(read), string(mntPort), block, -1)
		err = ioutil.WriteFile(ganeshaConfig, []byte(replaced), 0)
		if err != nil {
			return err
		}
	}

	return nil
}

func setGracePeriod(ganeshaConfig string, gracePeriod uint) error {
	if gracePeriod > 180 {
		return fmt.Errorf("grace period cannot be greater than 180")
	}

	newLine := fmt.Sprintf("Grace_Period = %d;", gracePeriod)

	re := regexp.MustCompile("Grace_Period = [0-9]+;")

	read, err := ioutil.ReadFile(ganeshaConfig)
	if err != nil {
		return err
	}

	oldLine := re.Find(read)

	var file *os.File
	if oldLine == nil {
		// Grace_Period line not there, append the whole NFSV4 block.
		file, err = os.OpenFile(ganeshaConfig, os.O_APPEND|os.O_WRONLY, 0600)
		if err != nil {
			return err
		}
		defer file.Close()

		block := "\nNFSV4\n{\n" +
			"\t" + newLine + "\n" +
			"}\n"

		if _, err = file.WriteString(block); err != nil {
			return err
		}
		file.Sync()
	} else {
		// Grace_Period line there, just replace it
		replaced := strings.Replace(string(read), string(oldLine), newLine, -1)
		err = ioutil.WriteFile(ganeshaConfig, []byte(replaced), 0)
		if err != nil {
			return err
		}
	}

	return nil
}

func setNumericOwners(ganeshaConfig string, numericOwners bool) error {
	newLine := fmt.Sprintf("Only_Numeric_Owners = %v;", numericOwners)

	re := regexp.MustCompile("Only_Numeric_Owners = (true|false);")

	read, err := ioutil.ReadFile(ganeshaConfig)
	if err != nil {
		return err
	}

	oldLine := re.Find(read)

	var file *os.File
	if oldLine == nil {
		// Only_Numeric_Owners line not there, append it after Grace_Period
		re := regexp.MustCompile("Grace_Period = [0-9]+;")

		gracePeriod := re.Find(read)

		block := string(gracePeriod) + "\n" +
			"\t" + newLine

		replaced := strings.Replace(string(read), string(gracePeriod), block, -1)
		err = ioutil.WriteFile(ganeshaConfig, []byte(replaced), 0)
		if err != nil {
			return err
		}
		file.Sync()
	} else {
		// Only_Numeric_Owners line there, just replace it
		replaced := strings.Replace(string(read), string(oldLine), newLine, -1)
		err = ioutil.WriteFile(ganeshaConfig, []byte(replaced), 0)
		if err != nil {
			return err
		}
	}

	return nil
}

// Stop stops the NFS server.
func Stop() {
	// /bin/dbus-send --system   --dest=org.ganesha.nfsd --type=method_call /org/ganesha/nfsd/admin org.ganesha.nfsd.admin.shutdown
}


cmd/nfs-provisioner/main.go
/*
Copyright 2016 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"strings"
	"time"

	"github.com/golang/glog"
	"github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner/pkg/server"
	vol "github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner/pkg/volume"
	"github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/controller"
	"k8s.io/apimachinery/pkg/util/validation"
	"k8s.io/apimachinery/pkg/util/validation/field"
	"k8s.io/apimachinery/pkg/util/wait"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)

var (
	provisioner      = flag.String("provisioner", "example.com/nfs", "Name of the provisioner. The provisioner will only provision volumes for claims that request a StorageClass with a provisioner field set equal to this name.")
	master           = flag.String("master", "", "Master URL to build a client config from. Either this or kubeconfig needs to be set if the provisioner is being run out of cluster.")
	kubeconfig       = flag.String("kubeconfig", "", "Absolute path to the kubeconfig file. Either this or master needs to be set if the provisioner is being run out of cluster.")
	runServer        = flag.Bool("run-server", true, "If the provisioner is responsible for running the NFS server, i.e. starting and stopping NFS Ganesha. Default true.")
	useGanesha       = flag.Bool("use-ganesha", true, "If the provisioner will create volumes using NFS Ganesha (D-Bus method calls) as opposed to using the kernel NFS server ('exportfs'). If run-server is true, this must be true. Default true.")
	gracePeriod      = flag.Uint("grace-period", 90, "NFS Ganesha grace period to use in seconds, from 0-180. If the server is not expected to survive restarts, i.e. it is running as a pod & its export directory is not persisted, this can be set to 0. Can only be set if both run-server and use-ganesha are true. Default 90.")
	enableXfsQuota   = flag.Bool("enable-xfs-quota", false, "If the provisioner will set xfs quotas for each volume it provisions. Requires that the directory it creates volumes in ('/export') is xfs mounted with option prjquota/pquota, and that it has the privilege to run xfs_quota. Default false.")
	serverHostname   = flag.String("server-hostname", "", "The hostname for the NFS server to export from. Only applicable when running out-of-cluster i.e. it can only be set if either master or kubeconfig are set. If unset, the first IP output by `hostname -i` is used.")
	exportSubnet     = flag.String("export-subnet", "*", "Subnet for NFS export to allow mount only from")
	maxExports       = flag.Int("max-exports", -1, "The maximum number of volumes to be exported by this provisioner. New claims will be ignored once this limit has been reached. A negative value is interpreted as 'unlimited'. Default -1.")
	fsidDevice       = flag.Bool("device-based-fsids", true, "If file system handles created by NFS Ganesha should be based on major/minor device IDs of the backing storage volume ('/export'). Default true.")
	leaderElection   = flag.Bool("leader-elect", false, "Start a leader election client and gain leadership before executing the main loop. Enable this when running replicated components for high availability. Default false.")
	disableIDMapping = flag.Bool("disable-id-mapping", false, "Configure NFS Ganesha to disable ID mapping for NFSv4 mounts. Can only be set if both run-server and use-ganesha are true. Default false.")
)

const (
	exportDir     = "/export"
	ganeshaLog    = "/export/ganesha.log"
	ganeshaPid    = "/var/run/ganesha.pid"
	ganeshaConfig = "/export/vfs.conf"
)

func main() {
	flag.Set("logtostderr", "true")
	flag.Parse()

	if errs := validateProvisioner(*provisioner, field.NewPath("provisioner")); len(errs) != 0 {
		glog.Fatalf("Invalid provisioner specified: %v", errs)
	}
	glog.Infof("Provisioner %s specified", *provisioner)

	if *runServer && !*useGanesha {
		glog.Fatalf("Invalid flags specified: if run-server is true, use-ganesha must also be true.")
	}

	if *useGanesha && *exportSubnet != "*" {
		glog.Warningf("If use-ganesha is true, there is no effect on export-subnet.")
	}

	if *gracePeriod != 90 && (!*runServer || !*useGanesha) {
		glog.Fatalf("Invalid flags specified: custom grace period can only be set if both run-server and use-ganesha are true.")
	} else if *gracePeriod > 180 && *runServer && *useGanesha {
		glog.Fatalf("Invalid flags specified: custom grace period must be in the range 0-180")
	}

	if *disableIDMapping && (!*runServer || !*useGanesha) {
		glog.Fatalf("Invalid flags specified: ID mapping can only be disabled if both run-server and use-ganesha are true.")
	}

	// Create the client according to whether we are running in or out-of-cluster
	outOfCluster := *master != "" || *kubeconfig != ""

	if !outOfCluster && *serverHostname != "" {
		glog.Fatalf("Invalid flags specified: if server-hostname is set, either master or kube-config must also be set.")
	}

	if *runServer {
		glog.Infof("Setting up NFS server!")
		err := server.Setup(ganeshaConfig, *gracePeriod, *fsidDevice, *disableIDMapping)
		if err != nil {
			glog.Fatalf("Error setting up NFS server: %v", err)
		}
		go func() {
			for {
				// This blocks until server exits (presumably due to an error)
				err = server.Run(ganeshaLog, ganeshaPid, ganeshaConfig)
				if err != nil {
					glog.Errorf("NFS server Exited Unexpectedly with err: %v", err)
				}

				// take a moment before trying to restart
				time.Sleep(time.Second)
			}
		}()
		// Wait for NFS server to come up before continuing provisioner process
		time.Sleep(5 * time.Second)
	}

	var config *rest.Config
	var err error
	if outOfCluster {
		config, err = clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
	} else {
		config, err = rest.InClusterConfig()
	}
	if err != nil {
		glog.Fatalf("Failed to create config: %v", err)
	}
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		glog.Fatalf("Failed to create client: %v", err)
	}

	// The controller needs to know what the server version is because out-of-tree
	// provisioners aren't officially supported until 1.5
	serverVersion, err := clientset.Discovery().ServerVersion()
	if err != nil {
		glog.Fatalf("Error getting server version: %v", err)
	}

	// Create the provisioner: it implements the Provisioner interface expected by
	// the controller
	nfsProvisioner := vol.NewNFSProvisioner(exportDir, clientset, outOfCluster, *useGanesha, ganeshaConfig, *enableXfsQuota, *serverHostname, *maxExports, *exportSubnet)

	// Start the provision controller which will dynamically provision NFS PVs
	pc := controller.NewProvisionController(
		clientset,
		*provisioner,
		nfsProvisioner,
		serverVersion.GitVersion,
		controller.LeaderElection(*leaderElection),
	)

	pc.Run(wait.NeverStop)
}

// validateProvisioner tests if provisioner is a valid qualified name.
// https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/apis/storage/validation/validation.go
func validateProvisioner(provisioner string, fldPath *field.Path) field.ErrorList {
	allErrs := field.ErrorList{}
	if len(provisioner) == 0 {
		allErrs = append(allErrs, field.Required(fldPath, provisioner))
	}
	if len(provisioner) > 0 {
		for _, msg := range validation.IsQualifiedName(strings.ToLower(provisioner)) {
			allErrs = append(allErrs, field.Invalid(fldPath, provisioner, msg))
		}
	}
	return allErrs
}
